tests/topotests/lib/bgp.py

```diff
@@ -5690,3 +5691,88 @@ def bgp_vpn_router_json_cmp_exact_filter(router, cmd, expected):
             json_output["routes"]["routeDistinguishers"].pop(rd)
 
     return topotest.json_cmp(json_output, expected, exact=True)
+
+
+def bgp_get_established_epoch(router, peer, vrf=None):
+    """
+    Get the established epoch for a peer
+    """
+    if vrf:
+        output = router.vtysh_cmd(
+            f"show bgp vrf {vrf} neighbor {peer} json", isjson=True
+        )
+    else:
+        output = router.vtysh_cmd(f"show bgp neighbor {peer} json", isjson=True)
+    assert peer in output, "peer not found"
+    peer_info = output[peer]
+    assert "bgpState" in peer_info, "peer state not found"
+    assert peer_info["bgpState"] == "Established", "peer not in Established state"
+    assert "bgpTimerUpEstablishedEpoch" in peer_info, "peer epoch not found"
+    return peer_info["bgpTimerUpEstablishedEpoch"]
+
+
+def bgp_check_established_epoch_differ(router, peer, last_established_epoch, vrf=None):
+    """
+    Check that the established epoch has changed
+    """
+    if vrf:
+        output = router.vtysh_cmd(
+            f"show bgp vrf {vrf} neighbor {peer} json", isjson=True
+        )
+    else:
+        output = router.vtysh_cmd(f"show bgp neighbor {peer} json", isjson=True)
+    assert peer in output, "peer not found"
+    peer_info = output[peer]
+    assert "bgpState" in peer_info, "peer state not found"
+
+    if peer_info["bgpState"] != "Established":
+        return "peer not in Established state"
+
+    assert "bgpTimerUpEstablishedEpoch" in peer_info, "peer epoch not found"
+
+    if peer_info["bgpTimerUpEstablishedEpoch"] == last_established_epoch:
+        return "peer epoch not changed"
+    return None
+
+
+def bgp_check_epoch_after_clear(router, peer, last_established_epoch, vrf=None):
+    """
+    Checking that the established epoch has changed and the peer is in Established state again after clear
+    Without this, the second session is cleared as well on slower systems (like CI)
+    """
+    test_func = partial(
+        bgp_check_established_epoch_differ,
+        router,
+        peer,
+        last_established_epoch,
+        vrf=vrf,
+    )
+    _, result = topotest.run_and_expect(test_func, None, count=20, wait=1)
+    assert (
+        result is None
+    ), "Established Epoch still the same after clear bgp for peer {}".format(peer)
+
+
+def bgp_check_wait_for_multipath_convergence(
+    router,
+    expected_paths=1,
+    vrf="vrf-101",
+    prefix_ip="10.0.101.1/32",
+    nexthop_ip="192.168.0.1",
+):
```

Comment:
I don't like the defaulting to test specific values in a lib

Reply:
perhaps don't put default values to raise a hard-error if those are missing?

otherwise, agreed.

Reply:
done.
with python, those parameters will be mandatory, except for:
- vrf: if not mentioned, the default routing table will be looked up
- expected_path: the only value which has a non None default value : set to 1.